compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -474,6 +474,10 @@ pub enum WellFormedLoc {
 #[derive(TypeVisitable, TypeFoldable)]
 pub struct ImplDerivedObligationCause<'tcx> {
     pub derived: DerivedObligationCause<'tcx>,
+    /// The `DefId` of the `impl` that gave rise to the `derived` obligation.
+    /// If the `derived` obligation arose from a trait alias, which conceptually has a synthetic impl,
+    /// then this will be the `DefId` of that trait alias. Care should therefore be taken to handle
+    /// that exceptional case where appropriate.
     pub impl_def_id: DefId,
```

Comment:
On reflection, maybe it'd be wise to rename this to `impl_or_alias_def_id` or similar?

Reply:
Yeah, sure go ahead

Reply:
I'll wait for #108115 to merge first, as it will conflict.